design/open/openapi-multipart-upload.md

```diff
@@ -0,0 +1,217 @@
+        description: relative to the branch
+        required: true
+        schema:
+          type: string
+      - in: query
+        name: parts
+        description: number of parts required to upload
+        schema:
+          type: integer
+          minimum: 1
+          maximum: 1000
+    post:
+      tags:
+        - staging
+      operationId: createMultipartUpload
+      summary: Initiate a multipart upload
+      description: Initiates a multipart upload and returns an upload ID with presigned URLs for each part. Part numbers starts with 1 (limit to 1000 parts). Each part minimum size is 5M.
+      responses:
+        200:
+          description: Multipart upload initiated
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/MultipartUpload"
+        # ... other error responses
+
+  /repositories/{repository}/branches/{branch}/staging/multipart/{uploadId}:
+      parameters:
+      - in: path
+        name: repository
+        required: true
+        schema:
+          type: string
+      - in: path
+        name: branch
+        required: true
+        schema:
+          type: string
+      - in: query
+        name: path
+        description: relative to the branch
+        required: true
+        schema:
+          type: string
+    put:
+      tags:
+        - staging
+      operationId: completeMultipartUpload
+      summary: Complete a multipart upload
+      description: Completes a multipart upload by assembling the uploaded parts.
+      requestBody:
+        content:
+          application/json:
+            schema:
+              $ref: "#/components/schemas/CompleteMultipartUpload"
+      responses:
+        200:
+          description: Multipart upload completed
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/ObjectStats"
+        # ... other error responses
+
+    delete:
+      tags:
+        - staging
+      operationId: abortMultipartUpload
+      summary: Abort a multipart upload
+      description: Aborts a multipart upload.
+      responses:
+        204:
+          description: Multipart upload aborted
+
+```
+
+
+**Schemas used by operations**
+
+```yaml=
+components:
+  schemas:
+    MultipartUpload:
+      type: object
+      properties:
+        upload_id:
+          type: string
+        locations:
+          type: array
+          items:
+            $ref: "#/components/schemas/StagingLocation"
+      required:
+        - upload_id
+        
+    UploadPart:
+      type: object
+      properties:
+        part_number:
+          type: integer
+          minimum: 1
+          maximum: 1000
+        etag:
+          type: string
+      required:
+        - part_number
+        - etag
+        
+    CompleteMultipartUpload:
+      type: object
+      properties:
+        parts:
+          type: array
+          items:
+            $ref: "#/components/schemas/UploadPart"
```

Comment:
Perhaps worth mentioning in the description that the object is created of these parts in the specified order -- **NOT** in the order in which the objects were uploaded, or in the order that `createMultipartUpload` returned the parts.

Reply:
Good point - the parts should be sorted by part number.

Reply:
Amusingly, the [S3 API](https://docs.aws.amazon.com/AmazonS3/latest/API/API_CompleteMultipartUpload.html) does not document such a requirement AFAICT?  But this is fine.

Reply:
Yea, you learn from the hard way.

```
Error Code: InvalidPartOrder

Description: The list of parts was not in ascending order. The parts list must be specified in order by part number.
HTTP Status Code: 400 Bad Request
```

Documented in https://docs.aws.amazon.com/cli/latest/reference/s3api/complete-multipart-upload.html